### DISCUSSION TITLE
Avoid nested injections; Inherit only the parser in custom filetypes

### DISCUSSION THREAD

**Discussion by goerz** [nvim-treesitter/nvim-treesitter#7373] [Q&A] [answered]
The [injection-queries for Julia](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/julia/injections.scm) that ship with `nvim-treesitter` enable the highlighting of docstrings as `markdown`. However, those docstrings then routinely contain Julia code blocks (for the signature, and possibly examples), and the nested inclusion of code blocks in `markdown` then kicks in. It makes things really confusing because then I can't easily tell whether I'm looking at the actual Julia code in the file, or some code example in a docstring.

Is there some way to disable nested injections, or some other workaround?

As far as workarounds go, I was thinking along the lines of changing the injection from `markdown` to a new custom `juliadocstring` filetype. I was hoping I could then link `juliadocstring` to `markdown` with `vim.treesitter.language.register("markdown", "juliadocstring")`, but then define override queries and syntax highlighting for the `juliadocstring` filetype. Alas, it appears that `vim.treesitter.language.register` doesn't just inherit the parser from `markdown`, but also all queries.

Does anyone have any other ideas, or a way to define a new custom filetype with an inherited parser but an independent `highlights.scm` / `injections.scm`?

**Comment by clason** [comment]
Override the Julia `injections` query in your own config to remove the Markdown injections.

**Reply by goerz** [reply]
I did that originally, but found out that I would prefer, in fact, to keep the injection of Markdown in the Julia file. What I want to disable is injections of code (Julia or otherwise) inside the injected Markdown. So, only one level of injection. Does that make sense?

**Reply by clason** [reply]
Sure, but that's not (reasonably) possible. There's (intentionally) no difference between a language whether it's injected or not.

**Reply by goerz** [reply]
Right, that's what I thought… so that's why I was wondering if there was a possible approach where I define a new language/filetype, (`juliadocstring`, or we could call it `markdown-injected`). That new filetype would inherit from `markdown`, but I could then modify it to, e.g., have an empty `injections.scm`, which would then disable the second-order inclusions.

I suppose I could fork the actual `markdown` treesitter parser to a new name and pretend this is a completely new language implemented from scratch. That seems a bit heavy-handed as a solution, though.

Do you have any suggestions or ideas on this front?

**Reply by clason** [reply, marked as answer]
> I suppose I could fork the actual markdown treesitter parser to a new name and pretend this is a completely new language implemented from scratch. That seems a bit heavy-handed as a solution, though.

Yes, that's what it would take, since the language name is embedded in the grammar.

My suggestion is to simply do without markdown in docstrings; when you're working on code, comment formatting is just a distraction.
